File: Games-Store-BE/src/main/java/com/gamesstorebe/entity/Product.java
package com.gamesstorebe.entity;

import com.fasterxml.jackson.annotation.*;
import jakarta.persistence.*;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Table;
import lombok.*;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Table(name = "products")
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;
    private String name;
    private String description;
    private Double price;
    private String releaseDate;
    private String status;

    private int downloads;

    @OneToOne(mappedBy = "product")
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    private ProductImage productImages;

    @ManyToOne
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    @JoinColumn(name = "categories_id")
    private Categories categories;

    @ManyToMany
    @JsonBackReference
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    @JoinTable(
            name = "products_featured",
            joinColumns = @JoinColumn(name = "product_id"),
            inverseJoinColumns = @JoinColumn(name = "featured_id")
    )
    private List<Features> productsFeatures;

    @ManyToOne
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    @JoinColumn(name = "developer_id")
    private Developers developer;


    @OneToOne(mappedBy = "product")
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    @JsonBackReference
    private ProductDiscount productDiscount;

    @ManyToMany(mappedBy = "products")
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    @JsonBackReference
    private List<Cart> carts;

    @ManyToMany(mappedBy = "productCheckout")
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    @JsonBackReference
    private List<Checkout> checkouts;
}

File: Games-Store-FE/src/app/core/models/product.ts
import { Discount } from "./discount";

export interface Product {
    id: number;
    name: string;
    description: string;
    price: number;
    releaseDate: Date;
    status: boolean;
    productImages: Image;
    category: Categories;
    productsFeatures: Features[];
    downloads: number;
    discount: Discount;
    categories: Categories;
    developer: Developers;

}

export interface Features {
    id: number;
    featureName: string;
}

export interface Categories {
    id: number;
    categoryName: string;
}

export interface Developers {
    id: number;
    developerName: string;
}

export interface Image {
    id: number;
    imageIcons: ImageIcon[];
    imageBanners: ImageBanner[];
    imageAvatars: ImageAvatar[];
}
export interface ImageBanner {
    id: number;
    name: string;
    path: string;
}

export interface ImageAvatar {
    id: number;
    name: string;
    path: string;
}

export interface ImageIcon {
    id: number;
    name: string;
    path: string;
}

export interface ProductDiscount {
    isHidden: boolean;
    id: number;
    product: Product;
    discount: Discount;
    // priceDiscount: number;
}
